### PR TITLE
feat(bwe) support setting the bandwidth from the client

### PR DESCRIPTION
**Pull request by mihhu**

### CODE REVIEW

File: react/features/base/ui/components/web/Input.tsx [hunk 1]
```diff
@@ -177,7 +181,7 @@ const Input = React.forwardRef<any, IProps>(({
                 {icon && <Icon
                     { ...(iconClick ? { tabIndex: 0 } : {}) }
                     className = { cx(styles.icon, iconClick && styles.iconClickable) }
-                    onClick = { iconClick }
+                    { ...(iconClick ? { onClick: iconClick } : {}) }
```
Comment (saghul) [inline react/features/base/ui/components/web/Input.tsx:184]: Why can't we let undefined be set here?

File: css/_base.scss [hunk 2]
```diff
@@ -193,3 +193,13 @@ form {
 .jitsi-icon svg path {
     fill: inherit !important;
 }
+
+input::-webkit-outer-spin-button,
```
Comment (saghul) [inline css/_base.scss:197]: Can't we do this in JSS instead?

File: config.js [hunk 3]
```diff
@@ -424,6 +424,8 @@ var config = {
 
     // Specify the settings for video quality optimizations on the client.
     // videoQuality: {
+    //    // Allow the setting of a custom bandwidth value from the UI.
+    //    assumeBandwidth: true,
```
Comment (saghul) [inline config.js:428]: Should we have this in the `testing` section while it mattures a bit?

File: react/features/base/conference/middleware.any.ts [hunk 4]
```diff
@@ -690,3 +696,33 @@ function _p2pStatusChanged(next: Function, action: AnyAction) {
...
+ * which is being dispatched in the specified {@code store}.
+ * @private
+ * @returns {Object} The value returned by {@code next(action)}.
+ */
+function _setAssumedBandwidthBps({ getState }: IStore, next: Function, action: AnyAction) {
+    const state = getState();
+    const conference = getCurrentConference(state);
+    const payload = Number(action.assumedBandwidthBps);
+
+    const assumedBandwidthBps = isNaN(payload) || payload < ASSUMED_BANDWIDTH_BPS
```
Comment (saghul) [inline react/features/base/conference/middleware.any.ts:719]: Can we pl rename `ASSUMED_BANDWIDTH_BPS` to `MIN_ASSUMED_BANDWIDTH_BPS` ? Makes it easier to read.

File: react/features/base/ui/components/web/Input.tsx [hunk 5]
```diff
@@ -16,8 +16,10 @@ interface IProps extends IInputProps {
     className?: string;
     iconClick?: () => void;
     id?: string;
+    max?: number;
```
Comment (saghul) [inline react/features/base/ui/components/web/Input.tsx:19]: Is this related to the value of the input? How about `minValue` ?

File: react/features/connection-indicator/components/web/BandwidthSettingsDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+const BandwidthSettingsDialog = () => {
+    const { classes } = useStyles();
+    const { t } = useTranslation();
+    const dispatch = useDispatch();
+    const [ showAssumedBandwidthInfo, setShowAssumedBandwidthInfo ] = useState(false);
+    const currentAssumedBandwidthBps = useSelector(
+        (state: IReduxState) => state['features/base/conference'].assumedBandwidthBps
+    );
+    const [ assumedBandwidthBps, setAssumedBandwidthBps ] = useState(
+        currentAssumedBandwidthBps === -1
```
Comment (saghul) [inline react/features/connection-indicator/components/web/BandwidthSettingsDialog.tsx:47]: Should we use the constant here?

File: react/features/connection-indicator/components/web/BandwidthSettingsDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    /**
+     * Persists the assumed bandwidth bps.
+     *
+     * @param {string} value - The key event to handle.
+     *
+     * @returns {void}
+     */
+    const onAssumedBandwidthBpsSave = useCallback(() => {
+        if (assumedBandwidthBps !== currentAssumedBandwidthBps) {
+            dispatch(saveAssumedBandwidthBps(Number(assumedBandwidthBps === '' ? -1 : assumedBandwidthBps)));
```
Comment (saghul) [inline react/features/connection-indicator/components/web/BandwidthSettingsDialog.tsx:73]: Ditto.

File: react/features/connection-indicator/components/web/BandwidthSettingsDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+    /**
+     * Validates the assumed bandwidth bps.
+     *
+     * @param {KeyboardEvent<any>} e - The key event to handle.
+     *
+     * @returns {void}
+     */
+    const onAssumedBandwidthBpsKeyPress = useCallback((e: KeyboardEvent<any>) => {
+        const isValid = (e.charCode !== 8 && e.charCode === 0) || (e.charCode >= 48 && e.charCode <= 57);
```
Comment (saghul) [inline react/features/connection-indicator/components/web/BandwidthSettingsDialog.tsx:85]: What are we checking for here?

File: react/features/connection-indicator/components/web/ConnectionIndicatorContent.tsx [hunk 9]
```diff
@@ -26,6 +27,8 @@ import AbstractConnectionIndicator, {
     INDICATOR_DISPLAY_THRESHOLD
 } from '../AbstractConnectionIndicator';
 
+import BandwidthSettingsDialog from './BandwidthSettingsDialog';
```
Comment (saghul) [inline react/features/connection-indicator/components/web/ConnectionIndicatorContent.tsx:30]: Is the connection indicator where we settled on displaying this? Wouldn't the performance dialog be more appropriate?

**Review by saghul — COMMENTED** [review]
Left some comments!

File: react/features/connection-indicator/components/web/ConnectionIndicatorContent.tsx (hunk 9 above)
Comment (mihhu) [reply react/features/connection-indicator/components/web/ConnectionIndicatorContent.tsx:30]: Yes, we wanted it to be a little bit more difficult to stumble upon.

File: react/features/connection-indicator/components/web/BandwidthSettingsDialog.tsx (hunk 6 above)
Comment (mihhu) [reply react/features/connection-indicator/components/web/BandwidthSettingsDialog.tsx:47]: That's right, I refactored it after quite some time and forgot about the constant

File: react/features/base/ui/components/web/Input.tsx (hunk 1 above)
Comment (mihhu) [reply react/features/base/ui/components/web/Input.tsx:184]: I was under the impression a warning would show up in the console. Will revert it

File: react/features/connection-indicator/components/web/BandwidthSettingsDialog.tsx (hunk 8 above)
Comment (mihhu) [reply react/features/connection-indicator/components/web/BandwidthSettingsDialog.tsx:85]: This should allow only digits and backspace

**Review by saghul — APPROVED** [review]
